frontend/gateway.go

```diff
@@ -237,51 +152,64 @@ func checkDiffMerge(client gwclient.Client) bool {
 }
 
 // copyForForward copies all the inputs and build opts from the initial request in order to forward to another frontend.
-func copyForForward(ctx context.Context, client gwclient.Client, req *gwclient.SolveRequest) error {
-	// Inputs are any additional build contexts or really any llb that the client sent along.
-	inputs, err := client.Inputs(ctx)
-	if err != nil {
-		return err
-	}
+func copyForForward(ctx context.Context, client gwclient.Client) solveRequestOpt {
+	return func(req *gwclient.SolveRequest) error {
+		// Inputs are any additional build contexts or really any llb that the client sent along.
+		inputs, err := client.Inputs(ctx)
+		if err != nil {
+			return err
+		}
 
-	if req.FrontendInputs == nil {
-		req.FrontendInputs = make(map[string]*pb.Definition, len(inputs))
-	}
+		if req.FrontendInputs == nil {
+			req.FrontendInputs = make(map[string]*pb.Definition, len(inputs))
+		}
+
+		for k, v := range inputs {
+			if _, ok := req.FrontendInputs[k]; ok {
+				// Do not overwrite existing inputs
+				continue
+			}
 
-	for k, v := range inputs {
-		if _, ok := req.FrontendInputs[k]; ok {
-			// Do not overwrite existing inputs
-			continue
+			def, err := v.Marshal(ctx)
+			if err != nil {
+				return errors.Wrap(err, "error marshaling frontend input")
+			}
+			req.FrontendInputs[k] = def.ToPB()
 		}
 
-		def, err := v.Marshal(ctx)
-		if err != nil {
-			return errors.Wrap(err, "error marshaling frontend input")
+		opts := client.BuildOpts().Opts
+		if req.FrontendOpt == nil {
+			req.FrontendOpt = make(map[string]string, len(opts))
 		}
-		req.FrontendInputs[k] = def.ToPB()
-	}
 
-	opts := client.BuildOpts().Opts
-	if req.FrontendOpt == nil {
-		req.FrontendOpt = make(map[string]string, len(opts))
-	}
+		for k, v := range opts {
 
-	for k, v := range opts {
+			if k == "filename" || k == "dockerfilekey" || k == "target" {
+				// These are some well-known keys that the dockerfile frontend uses
+				// which we'll be overriding with our own values (as needed) in the
+				// caller.
+				// Basically there should not be a need, nor is it desirable, to forward these along.
+				continue
+			}
 
-		if k == "filename" || k == "dockerfilekey" || k == "target" {
-			// These are some well-known keys that the dockerfile frontend uses
-			// which we'll be overriding with our own values (as needed) in the
-			// caller.
-			// Basically there should not be a need, nor is it desirable, to forward these along.
-			continue
+			if _, ok := req.FrontendOpt[k]; ok {
+				// Do not overwrite existing opts
+				continue
+			}
+			req.FrontendOpt[k] = v
 		}
 
-		if _, ok := req.FrontendOpt[k]; ok {
-			// Do not overwrite existing opts
-			continue
-		}
-		req.FrontendOpt[k] = v
+		return nil
 	}
+}
+
+const keyTargetOpt = "dalec.target"
```

Comment:
Can we change the name of this constant to `topLevelTarget` or something that really indicates it's supposed to be the top-level path component? I found it very easy to mix up `keyTargetOpt` and `keyTarget` when reading the code.

Reply:
That makes sense, done.